godot-core/src/task/futures.rs

```diff
@@ -308,8 +334,188 @@ impl<C: WithBaseField, R: ParamTuple + Sync + Send> IntoFuture for &TypedSignal<
     }
 }
 
+/// Convert a value into a type that is [`Send`] at compile-time while the value might not be.
+///
+/// This allows to turn any implementor into a type that is `Send`, but requires to also implement [`AssertSafeSend`] as well.
+/// The later trait will verify if a value can actually be sent between threads at runtime.
+#[expect(private_bounds)]
+pub trait IntoMaybeSend: Sealed {
+    type Target: Send + AssertSafeSend<Target = Self>;
+
+    fn into_maybe_send(self) -> Self::Target;
+}
+
+/// Assert that it was safe to send the value to the current thread.
+///
+/// # Safety
+/// The implementor has to guarantee that `assert_safe_send` panics if the value has been sent between threads while being `!Send`.
+#[expect(private_bounds)]
+pub unsafe trait AssertSafeSend: Send + Sealed {
+    type Target;
+
+    fn assert_safe_send(self) -> Self::Target;
+}
+
+trait Sealed {}
+
+pub struct ThreadCrosser<T> {
+    value: T,
+    thread_id: ThreadId,
+}
+
+unsafe impl<T> Send for ThreadCrosser<T> {}
+unsafe impl<T> Sync for ThreadCrosser<T> {}
+
+impl<T> ThreadCrosser<T> {
+    fn new(value: T) -> Self {
+        Self {
+            value,
+            thread_id: std::thread::current().id(),
+        }
+    }
+
+    fn extract(self) -> T {
+        assert_eq!(self.thread_id, std::thread::current().id());
+
+        self.value
+    }
+}
```

Comment:
How can this assertion ever be true if a thread is crossed? Isn't the purpose of `ThreadCrosser` to call `extract` on a different thread than `new`? 

If not, the name is misleading, and should be something like `ThreadConfined`...

Reply:
yes, I like `ThreadConfined` much better.